src/mame/machine/k573fpga.cpp

```diff
@@ -13,6 +13,7 @@
 
 k573fpga_device::k573fpga_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
 	device_t(mconfig, KONAMI_573_DIGITAL_FPGA, tag, owner, clock),
+	ram(*this, ":k573dio:ram"),
```

Comment:
Once again, this makes assumptions about the system hosting it, which is something we try to avoid.  The preferred way to do this kind of connection would be…

Construct it here as:
```
ram(*this, finder_base::DUMMY_TAG),
```

In the header provide a configuration method for setting the tag:
```
template <typename T> void set_ram(T &&tag) { ram.set_tag(std::forward<T>(tag)); }
```

Then when you instantiate it in the DIO device, you should be able to do this:
```
KONAMI_573_DIGITAL_FPGA(config, k573fpga);
k573fpga->set_ram(ram);
```

That way, the literal tag for the RAM is only in one place, you should get a validity check error on start if the RAM isn’t found, and it’s not relying on where any of the devices are in relation to each other.

Reply:
I understand what you were asking for now. I've implemented the code but there's an error and I'm not comfortable making the (seemingly?) required change without approval.

`k573fpga->set_ram(ram);` won't compile since `set_ram` wants a tag, not the actual `memory_share_creator`. My assumption is that I would need to make this code `k573fpga->set_ram(ram.finder_tag());` to actually get the tag from the `memory_share_creator`. However, it doesn't seem like it's actually possible to get the tag from `memory_share_creator` due to access level issues. 

`memory_share_creator` is declared as `class memory_share_creator : finder_base` which is making everything inside `finder_base` private. Changing it to `class memory_share_creator : public finder_base` allows me to use `finder_tag()` from `finder_base` but I don't have a deep enough understanding of the codebase to really understand if there are any unwanted implications involved with exposing `finder_base` functions here.

Or I could just use a const string for the tag in k573dio.

Please let me know what you think about the proposed change to `memory_share_creator`.

Reply:
It doesn’t require a literal tag, it can take on the target of another object finder.  Notice the overload `finder_base::set_tag(finder_base const &)`: https://github.com/mamedev/mame/blob/mame0227/src/emu/devfind.h#L304

We use this pattern pretty heavily to avoid repeating literal tags.

Private inheritance for `memory_share_creator` and `memory_bank_creator` is indeed an issue.  I think this is just an oversight – it should be fine to just change them both to use public inheritance.  Unfortunately it will trigger a recompilation of everything, since it’s part of the precompiled header.

If that change offends @galibert, I’ll take the heat.

Reply:
You're right, I just tried doing it the way you wrote in the code snippet and it's compiling just fine. I guess the private inheritance was what was causing it to not compile in the first place and I misinterpreted the error message. 

I'll make the change to `memory_share_creator` as well as `memory_bank_creator` then.

Thanks.